### PR TITLE
control flow concept updated

### LINKED ISSUE #39: Write the "Control Flow" concept
Official docs: 
https://book.cairo-lang.org/ch02-05-control-flow.html
https://book.cairo-lang.org/ch06-03-concise-control-flow-with-if-let-and-while-let.html

The idea is to have a short and to-the-point article about the concept. Avoid copy/pasting everything from The Cairo Book; the Book should be used as a more in-depth learning resource for students after they finish the concept on Exercism (or they can use it in parallel). 

---
Exercism Concepts docs:
https://exercism.org/docs/building/tracks/concepts

Use the following Rust's concept as inspiration for the format:
https://github.com/exercism/rust/blob/main/concepts/structs/about.md

### LINKED ISSUE #118: Introduce the concept of "gas" in some of the existing concepts
Students will encounter "gas used" in console logs, and might not understand what it means. We need to explain the concept asap, with a short description that's enough for them to continue learning.

### LINKED ISSUE #123: Add the "if let" and "while let" syntax description to the existing "Enums" concept
Official docs: 
https://book.cairo-lang.org/ch06-03-concise-control-flow-with-if-let-and-while-let.html

### PR DESCRIPTION
**Pull request by Falilah**
Closes #39
Closes #123 
Closes #118

### CODE REVIEW

File: concepts/control-flow/about.md [hunk 1]
```diff
@@ -1 +1,250 @@
...
+}
+```
+
+### Looping Through a Collection with `for`
+
+`for` Iterates over each item in a collection, executing code for each item.
+
+```rust
+fn main() {
+    let a: [felt; 5] = [10, 20, 30, 40, 50].span();
```
Comment (0xNeshi) [inline concepts/control-flow/about.md:193]:
```suggestion
    let a = [10, 20, 30, 40, 50].span();
```

File: concepts/control-flow/about.md [hunk 2]
```diff
@@ -1 +1,250 @@
...
+    for element: felt in a {
```
Comment (0xNeshi) [inline concepts/control-flow/about.md:194]:
```suggestion
    for element in a {
```

File: concepts/control-flow/introduction.md [hunk 3]
```diff
@@ -1 +1,3 @@
 # Introduction
+
+Control flow in Cairo, through constructs like `if` expressions and `loops`, allows developers to direct the execution of their programs based on conditional logic or repetitive tasks. If expressions enable branching based on whether conditions evaluate to `true` or `false`, while loops (`loop`, `while`, `for`) facilitate repeated execution of code blocks until specific criteria are met. These constructs are crucial for implementing logic, iteration, and handling multiple conditions concisely, making them essential tools for writing efficient and dynamic code in Cairo.
```
Comment (0xNeshi) [inline concepts/control-flow/introduction.md:3]:
```suggestion
Control flow in Cairo, through constructs like `if` expressions and loops, allows developers to direct the execution of their programs based on conditional logic or repetitive tasks. `if` expressions enable branching based on whether conditions evaluate to `true` or `false`, while loops (`loop`, `while`, `for`) facilitate repeated execution of code blocks until specific criteria are met. These constructs are crucial for implementing logic, iteration, and handling multiple conditions concisely, making them essential tools for writing efficient and dynamic code in Cairo.
```
Very nice intro

File: concepts/control-flow/.meta/config.json [hunk 4]
```diff
@@ -1,7 +1,5 @@
 {
-    "blurb": "<todo>",
-    "authors": [
-        "0xNeshi"
-    ],
-    "contributors": []
+  "blurb": "Control flow in Cairo uses if expressions to conditionally execute code and loops to repeat code while a condition is true.",
```
Comment (0xNeshi) [inline concepts/control-flow/.meta/config.json:2]:
```suggestion
  "blurb": "Control flow in Cairo uses `if` expressions to conditionally execute code and loops to run code repeatedly.",
```

File: concepts/control-flow/.meta/config.json [hunk 5]
```diff
@@ -1,7 +1,5 @@
...
+  "authors": ["misicnenad", "Falilah"],
```
Comment (0xNeshi) [inline concepts/control-flow/.meta/config.json:3]:
```suggestion
  "authors": ["Falilah"],
```
Most of the changes I recommended are nits, so no need to list me as a co-author. The honor is all yours :)

File: concepts/control-flow/.meta/config.json [hunk 6]
```diff
@@ -1,7 +1,5 @@
...
+  "contributors": []
```
Comment (0xNeshi) [inline concepts/control-flow/.meta/config.json:4]:
```suggestion
  "contributors": ["0xNeshi"]
```
I changed my GH username

**Review by 0xNeshi — CHANGES_REQUESTED** [review]
Really good job! Most of the comments are nits.

File: concepts/control-flow/about.md [hunk 7]
```diff
@@ -1 +1,250 @@
...
+        println!("the value is: {}", element);
+    }
+}
+```
+
+## Concise Control Flow with `if let` and `while let`
```
[the remark this answers is not in the document]
Comment (Falilah) [reply concepts/control-flow/about.md:200]: to get this point right, should it be be moved to `enum` concept or copied to` enum` concept ?

[the remark this answers is not in the document]
Comment (0xNeshi) [reply concepts/control-flow/about.md:200]: They should be moved, i.e. they should only appear in the "Enums" concept.
Created a new issue for this https://github.com/exercism/cairo/issues/123

File: concepts/control-flow/about.md [hunk 8]
```diff
@@ -1,34 +1,29 @@
...
 
-- Conditionally running code based on a `true` condition.
-- Repeating code while a condition remains `true`.
+- **`If` expressions** conditionally run code based on whether a condition evaluates to true.
+- **Loops** (such as `loop`, `while`, and `for`) repeatedly execute code while a specified condition remains true or until a breaking condition is met.
 
 These constructs are fundamental for implementing logic and iteration in Cairo.
 
-## If Expressions
+## `If` Expressions
```
Comment (0xNeshi) [inline concepts/control-flow/about.md:11]:
```suggestion
## `if` Expressions
```

File: concepts/control-flow/about.md [hunk 9]
```diff
@@ -1,34 +1,29 @@
...
 
-If expressions in Cairo allow you to control the flow of execution based on conditional logic. They follow a structure where code is executed depending on whether a condition evaluates to `true` or `false`.
-
-## Basic If Expression
+`If` expressions in Cairo allow you to control the flow of execution based on conditional logic. They follow a structure where code is executed depending on whether a boolean condition evaluates to `true` or `false`.
```
Comment (0xNeshi) [inline concepts/control-flow/about.md:13]:
```suggestion
`if` expressions in Cairo allow you to control the flow of execution based on conditional logic. They follow a structure where code is executed depending on whether a boolean condition evaluates to `true` or `false`.
```

File: concepts/control-flow/about.md [hunk 10]
```diff
@@ -110,9 +99,11 @@ fn main() {
 }
 ```
 
-Executes code indefinitely until stopped manually or via `break`.
+Executing this code will print `"again!"` indefinitely in the console until either the program runs out of gas or we stop it manually.
+
+> Note: Cairo prevents us from running program with infinite loops by including a gas meter. The gas meter is a mechanism that limits the amount of computation that can be done in a program. Gas is a unit of measurement that expresses the computation cost of an instruction. When the gas meter runs out, the program will stop.
```
Comment (0xNeshi) [inline concepts/control-flow/about.md:104]:
```suggestion
> Note: Cairo prevents us from running a program with infinite loops by including a gas meter. The gas meter is a mechanism that limits the amount of computation that can be done in a program. Gas is a unit of measurement that expresses the computation cost of an instruction. When the gas meter runs out, the program will stop.
```

File: concepts/control-flow/about.md [hunk 11]
```diff
@@ -148,34 +139,32 @@ fn main() {
 
 Executing this program will not print the value of `i` when it is equal to `5`.
 
-## Returning Values from loops
+#### Returning Values from loops
```
Comment (0xNeshi) [inline concepts/control-flow/about.md:142]:
```suggestion
#### Returning Values from Loops
```

File: concepts/control-flow/introduction.md [hunk 12]
```diff
@@ -1,3 +1,3 @@
 # Introduction
 
-Control flow in Cairo, through constructs like `if` expressions and `loops`, allows developers to direct the execution of their programs based on conditional logic or repetitive tasks. If expressions enable branching based on whether conditions evaluate to `true` or `false`, while loops (`loop`, `while`, `for`) facilitate repeated execution of code blocks until specific criteria are met. These constructs are crucial for implementing logic, iteration, and handling multiple conditions concisely, making them essential tools for writing efficient and dynamic code in Cairo.
+Control flow in Cairo, through constructs like `if` expressions and loops, allows developers to direct the execution of their programs based on conditional logic or repetitive tasks. `If` expressions enable branching based on whether conditions evaluate to `true` or `false`, while loops (`loop`, `while`, `for`) facilitate repeated execution of code blocks until specific criteria are met. These constructs are crucial for implementing logic, iteration, and handling multiple conditions concisely, making them essential tools for writing efficient and dynamic code in Cairo.
```
Comment (0xNeshi) [inline concepts/control-flow/introduction.md:3]:
```suggestion
Control flow in Cairo, through constructs like `if` expressions and loops, allows developers to direct the execution of their programs based on conditional logic or repetitive tasks. `if` expressions enable branching based on whether conditions evaluate to `true` or `false`, while loops (`loop`, `while`, `for`) facilitate repeated execution of code blocks until specific criteria are met. These constructs are crucial for implementing logic, iteration, and handling multiple conditions concisely, making them essential tools for writing efficient and dynamic code in Cairo.
```

File: concepts/enums/about.md [hunk 13]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
 }
 ```
 
+## Concise Control Flow with `if let` and `while let`
+
+`if let`
```
Comment (0xNeshi) [inline concepts/enums/about.md:143]:
```suggestion
```

File: concepts/enums/about.md [hunk 14]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
 }
 ```
 
+## Concise Control Flow with `if let` and `while let`
+
```
Comment (0xNeshi) [inline concepts/enums/about.md:142]:
```suggestion
```

File: concepts/enums/about.md [hunk 15]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+`if let`
+
+The `if let` syntax combines `if` and `let` to handle specific pattern matches more concisely.
+
+Using `match` to handle an Option:
```
Comment (0xNeshi) [inline concepts/enums/about.md:147]:
```suggestion
Using `match` to handle an `Option`:
```

File: concepts/enums/about.md [hunk 16]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+
+```rust
+let config_max: Option<felt> = Option::Some(5);
```
Comment (0xNeshi) [inline concepts/enums/about.md:150]:
```suggestion
let config_max: Option<felt252> = Option::Some(5);
```

File: concepts/enums/about.md [hunk 17]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+match config_max {
+    Option::Some(max: felt) => println!("The maximum is configured to be {}", max),
```
Comment (0xNeshi) [inline concepts/enums/about.md:152]:
```suggestion
    Option::Some(max) => println!("The maximum is configured to be {}", max),
```

File: concepts/enums/about.md [hunk 18]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+    _ => (),
+}
+```
+
+Using `if let` for the same logic:
+
+```rust
+let number: Option<felt>  = Option::Some(5);
```
Comment (0xNeshi) [inline concepts/enums/about.md:160]:
```suggestion
let number: Option<felt252>  = Option::Some(5);
```

File: concepts/enums/about.md [hunk 19]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+if let Option::Some(max: felt) = number {
```
Comment (0xNeshi) [inline concepts/enums/about.md:161]:
```suggestion
if let Option::Some(max) = number {
```

File: concepts/enums/about.md [hunk 20]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+    println!("The maximum is configured to be {}", max);
+}
+```
+
+You can also include else with if let:
```
Comment (0xNeshi) [inline concepts/enums/about.md:166]:
```suggestion
You can also include an `else` with `if let`:
```

File: concepts/enums/about.md [hunk 21]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+
+```rust
+let coin: Coin = Coin::Quarter;
+let mut count: felt = 0;
```
Comment (0xNeshi) [inline concepts/enums/about.md:170]:
```suggestion
```
makes little sense given the example

File: concepts/enums/about.md [hunk 22]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+if let Coin::Quarter = coin {
+    println!("You got a quarter!");
+} else {
+    count += 1;
```
Comment (0xNeshi) [inline concepts/enums/about.md:174]:
```suggestion
    println!("You got some other coin!");
```

File: concepts/enums/about.md [hunk 23]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+}
+```
+
+`while let`:
```
Comment (0xNeshi) [inline concepts/enums/about.md:178]:
```suggestion
```

File: concepts/enums/about.md [hunk 24]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+
+The `while let` syntax loops over values, executing a block for each matching pattern.
```
Comment (0xNeshi) [inline concepts/enums/about.md:180]:
```suggestion
The `while let` syntax loops over `Option` values, executing a block for each matching pattern.
```

File: concepts/enums/about.md [hunk 25]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+
+```rust
+fn main() {
+    let mut arr: Array<felt> = array![1, 2, 3, 4, 5, 6, 7, 8, 9];
```
Comment (0xNeshi) [inline concepts/enums/about.md:184]:
```suggestion
    let mut arr: Array<felt252> = array![1, 2, 3, 4, 5, 6, 7, 8, 9];
```

File: concepts/enums/about.md [hunk 26]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+    let mut sum: felt = 0;
```
Comment (0xNeshi) [inline concepts/enums/about.md:185]:
```suggestion
    let mut sum: felt252 = 0;
```

File: concepts/enums/about.md [hunk 27]
```diff
@@ -138,4 +138,56 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+    while let Option::Some(value: felt) = arr.pop_front() {
```
Comment (0xNeshi) [inline concepts/enums/about.md:186]:
```suggestion
    while let Option::Some(value) = arr.pop_front() {
```

File: concepts/control-flow/about.md [hunk 28]
```diff
@@ -1 +1,187 @@
 # Control Flow
+
+`If` expressions and loops control the flow of execution in Cairo by:
```
Comment (0xNeshi) [inline concepts/control-flow/about.md:3]:
```suggestion
`if` expressions and loops control the flow of execution in Cairo by:
```

File: concepts/control-flow/about.md [hunk 29]
```diff
@@ -1 +1,187 @@
...
+`
+
+- **`If` expressions** conditionally run code based on whether a condition evaluates to true.
```
Comment (0xNeshi) [inline concepts/control-flow/about.md:6]:
```suggestion
- **`if` expressions** conditionally run code based on whether a condition evaluates to true.
```

**Review by 0xNeshi — APPROVED** [review]

File: concepts/enums/about.md [hunk 30]
```diff
@@ -138,4 +138,52 @@ fn find_value(mut arr: Span<felt252>, value: felt252) -> Option<usize> {
...
+
+```rust
+let coin: Coin = Coin::Quarter;
+if let Coin::Quarter = coin {
+    println!("You got a quarter!");
+} else {
+    println!("You got some other coin!");
+}
+```
+
```
Comment (0xNeshi) [inline concepts/enums/about.md:174]:
```suggestion
```

**Comment by 0xNeshi** [comment]
I applied some nits to avoid bugging you about them.

Great work! :blush:

**Comment by Falilah** [comment]
> I applied some nits to avoid bugging you about them.
> 
> Great work! 😊
Thank you ✅

**Pull request merged**